lib/nerves_hub_link/update_manager.ex

```diff
@@ -145,22 +163,55 @@ defmodule NervesHubLink.UpdateManager do
   end
 
   # messages from Downloader
-  def handle_info({:download, :complete}, state) do
+  def handle_info({:download, :complete, _fwup_public_keys}, state) do
     Logger.info("[NervesHubLink] Firmware Download complete")
-    {:noreply, %State{state | download: nil}}
+
+    {:noreply, %State{state | status: :idle, retrying: false, previous_update: :complete}}
   end
 
-  def handle_info({:download, {:error, reason}}, state) do
-    Logger.error("[NervesHubLink] Nonfatal HTTP download error: #{inspect(reason)}")
+  def handle_info(
+        {:download, {:error, %Mint.HTTPError{reason: {:http_error, 401}}}, _fwup_public_keys},
+        %State{retrying: true} = state
+      ) do
+    Logger.error("[NervesHubLink] Firmware download error: 401 after retry")
+    {:noreply, %State{state | status: :idle, retrying: false, previous_update: :failed}}
+  end
+
+  def handle_info(
+        {:download, {:error, %Mint.HTTPError{reason: {:http_error, 401}}}, fwup_public_keys},
+        state
+      ) do
+    Logger.error(
+      "[NervesHubLink] Firmware download URL signature may have expired. Requesting new URL and retrying."
+    )
+
+    GenServer.stop(Fwup.Stream)
+
+    update_info = Socket.check_update_available()
```

Comment:
It's prob worth making this async. With this, you have a hard limit of 5 seconds before the genserver call times out. If the request to the server and resolution of the update takes 5.1 seconds, then this all fails

Switching to async means you fire a message to the server, it can take its time getting back, and you know how to respond to the next update message when received.

If we're really worried about it, maybe we have an UpdateRetry that fails after a few minutes to indicate the server never told it a new URL? (Or just a log message)

Reply:
I agree that async would be better.

It will take some reworking, including in the Hub Web code.